subprojects/language-groovy/src/main/java/org/gradle/api/tasks/compile/GroovyCompile.java

```diff
@@ -303,11 +307,50 @@ private GroovyJavaJointCompileSpec createSpec() {
             GFileUtils.mkdirs(dir);
             spec.getGroovyCompileOptions().setStubDir(dir);
         }
-        spec.getCompileOptions().getForkOptions().setExecutable(Jvm.current().getJavaExecutable().getAbsolutePath());
+
+        if (javaLauncher.isPresent()) {
+            applyToolchain(spec.getCompileOptions().getForkOptions());
+        } else {
+            spec.getCompileOptions().getForkOptions().setExecutable(Jvm.current().getJavaExecutable().getAbsolutePath());
+        }
 
         return spec;
     }
 
+    private void configureCompatibilityOptions(DefaultGroovyJavaJointCompileSpec spec) {
+        JavaInstallationMetadata toolchain = getToolchain();
+        if (toolchain != null) {
+            boolean isSourceOrTargetConfigured = false;
+            if (super.getSourceCompatibility() != null) {
+                spec.setSourceCompatibility(getSourceCompatibility());
+                isSourceOrTargetConfigured = true;
+            }
+            if (super.getTargetCompatibility() != null) {
+                spec.setTargetCompatibility(getTargetCompatibility());
+                isSourceOrTargetConfigured = true;
+            }
+            if (!isSourceOrTargetConfigured) {
+                String languageVersion = toolchain.getLanguageVersion().toString();
+                spec.setSourceCompatibility(languageVersion);
+                spec.setTargetCompatibility(languageVersion);
+            }
+        } else {
+            spec.setSourceCompatibility(getSourceCompatibility());
+            spec.setTargetCompatibility(getTargetCompatibility());
+        }
+    }
+
+    private void applyToolchain(ForkOptions forkOptions) {
+        forkOptions.setExecutable(javaLauncher.get().getExecutablePath().getAsFile().getAbsolutePath());
```

Comment:
Setting just the executable will disable the cache. I think you either want java home or disable this (questionable) cache flag:

https://github.com/gradle/gradle/blob/41ccab532c7c2724796c996ed8884155982ee5a1/subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java#L117

Reply:
No that does not apply. What this code looks at is the `CompileOptions`, here the value is set on the spec.
Now looking at the code, I extracted a method the wrong way. Going to rework that a bit to make it clearer.

Reply:
And btw this shows how this is really busted in a way: if you attempt to set the executable, you will disable caching but then the set value will be ignored and replaced ...